internal/controller/controllers/cluster_controller.go

```diff
@@ -171,17 +228,121 @@ func (r *ClusterReconciler) updateStatus(ctx context.Context, cluster *adiiov1al
+	update := false
+
+	params := &models.ClusterUpdateParams{}
+
+	spec := cluster.Spec
+
+	updateString := func(new, old string, target **string) {
+		if new != old {
+			*target = swag.String(new)
+			update = true
+			r.Log.Infof("updating %s != %s", old, new)
+		}
+	}
+
+	updateString(spec.Name, c.Name, &params.Name)
+
+	if spec.OpenshiftVersion != c.OpenshiftVersion {
+		return false, ctrl.Result{}, errors.Errorf("Openshift version cannot be updated")
+	}
+
+	updateString(spec.BaseDNSDomain, c.BaseDNSDomain, &params.BaseDNSDomain)
+	updateString(spec.ClusterNetworkCidr, c.ClusterNetworkCidr, &params.ClusterNetworkCidr)
+
+	if spec.ClusterNetworkHostPrefix != c.ClusterNetworkHostPrefix {
+		params.ClusterNetworkHostPrefix = swag.Int64(spec.ClusterNetworkHostPrefix)
+		update = true
+	}
+
+	updateString(spec.ServiceNetworkCidr, c.ServiceNetworkCidr, &params.ServiceNetworkCidr)
+	updateString(spec.APIVip, c.APIVip, &params.APIVip)
+	updateString(spec.APIVipDNSName, swag.StringValue(c.APIVipDNSName), &params.APIVipDNSName)
+	updateString(spec.IngressVip, c.IngressVip, &params.IngressVip)
+	updateString(spec.MachineNetworkCidr, c.MachineNetworkCidr, &params.MachineNetworkCidr)
+	updateString(spec.SSHPublicKey, c.SSHPublicKey, &params.SSHPublicKey)
+
+	if spec.VIPDhcpAllocation != swag.BoolValue(c.VipDhcpAllocation) {
+		params.VipDhcpAllocation = swag.Bool(spec.VIPDhcpAllocation)
+	}
+
+	updateString(spec.HTTPProxy, c.HTTPProxy, &params.HTTPProxy)
+	updateString(spec.HTTPSProxy, c.HTTPSProxy, &params.HTTPSProxy)
+	updateString(spec.NoProxy, c.NoProxy, &params.NoProxy)
+
+	if spec.UserManagedNetworking != swag.BoolValue(c.UserManagedNetworking) {
+		params.UserManagedNetworking = swag.Bool(spec.UserManagedNetworking)
+	}
+
+	updateString(spec.AdditionalNtpSource, c.AdditionalNtpSource, &params.AdditionalNtpSource)
+
+	// TODO: handle InstallConfigOverrides
+
+	data, err := r.getPullSecret(ctx, spec.PullSecretRef.Name, spec.PullSecretRef.Namespace)
+	if err != nil {
+		return false, ctrl.Result{}, errors.Wrap(err, "failed to get pull secret for update")
+	}
+	updateString(data, c.PullSecret, &params.PullSecret)
+
+	if !update {
+		return update, ctrl.Result{}, nil
+	}
+
+	updatedCluster, err := r.Installer.UpdateClusterInternal(ctx, installer.UpdateClusterParams{
+		ClusterUpdateParams: params,
+		ClusterID:           *c.ID,
+	})
+
+	if err != nil {
+		return update, ctrl.Result{}, errors.Wrap(err, "failed to update cluster")
+	}
+
+	r.Log.Infof("Updated cluster %s %s", cluster.Name, cluster.Namespace)
+	reply, err := r.updateState(ctx, cluster, updatedCluster)
+	return update, reply, err
+
 }
 
 func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	mapSecretToCluster := handler.ToRequestsFunc(
+		func(a handler.MapObject) []reconcile.Request {
+			clusters := &adiiov1alpha1.ClusterList{}
+			if err := r.List(context.Background(), clusters); err != nil {
```

Comment:
In order to avoid iteration on all the cluster I **think** something like this will be more optimized:

```
			options := &client.ListOptions{
				Namespace: a.Meta.GetNamespace(),
				FieldSelector: fields.SelectorFromSet(fields.Set{
					"spec.PullSecretRef.Name": a.Meta.GetName(),
					"spec.PullSecretRef.Namespace": a.Meta.GetNamespace(),
				}),
			}
			if err := r.List(context.Background(), clusters, options); err != nil {
				return []reconcile.Request{}
			}
```

Reply:
* all the cluster resources will be cached in-memory anyway, so there's no network activity to worry about
* pull secret changes should be rare
* it's reasonable to expect that in many cases, there will only be one pull secret, or a small number, so a change is likely to apply to a large portion of clusters anyway

The last time I looked at using a field selector like this, it required also setting up an index on the field in advance, and turned out to be fairly complex. The situation may have improved since then, but it's worth keeping in mind.

Reply:
> It's probably better to use the `SyncPeriod` field on this Options resource: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/manager#Options
> 
> It causes all controllers for that manager to do the same, but that's probably what you want at this point anyway.

This configuration is on a managers level. i need to set it only for clusters resource, and it's a temporary solution until we have notifications from backend, i tried to look for something on a controller level and didn't find anything the specify for a specific controller, https://github.com/operator-framework/operator-sdk/issues/3086. 
I hope that we will be able to get backend notifications  as fast as possible and this code will be removed.

Reply:
Looks like @mhrivnak is right, i tired with multiple combinations and it always return an empty list.
So probably some indexing is required. so for now i will leave it as is.